Add an off drawer canvas region for slide-in menu bar

alice-0-kim opened the pull request:
Initial pull request to `develop` branch.

alice-0-kim commented:
Hi Joel, no worries. Thank you for the time you've spent on the code review. Please let me know if you have additional comments.

Thank you!

joelpittet commented on template.php:
```diff
@@ -479,7 +519,7 @@ function megatron_megatron_btn_dropdown($variables) {
   // Finish markup
   $output .= '
-  <div class="ubc7-arrow down-arrow"></div>
+  <div class="ubc7-arrow right-arrow"></div>
```
These may break BC, could we use maybe a different theme?

alice-0-kim replied:
Yes, it should be down-arrow. Thank you for bringing this up.

joelpittet commented on template.php:
```diff
@@ -919,7 +959,7 @@ function megatron_megatron_links($variables) {
           $link['html'] = TRUE;
           $output .=  '<div class="btn-group">' .l($link['title'], $link['href'], $link);
-          $output .=  '<button class="btn dropdown-toggle" data-toggle="dropdown"><span class="ubc7-arrow blue down-arrow"></span></button>';
+          $output .=  '<button class="btn dropdown-toggle" data-toggle="dropdown"><span class="ubc7-arrow white right-arrow"></span></button>';
```
Same here with blue and white change could have some negative changes to existing sites?

alice-0-kim commented:
@joelpittet Thank you for bringing up this issue. I've committed the changes so that the arrow on the main navigation menu a blue arrow pointing down while the arrow on the drawer region is a white arrow pointing to the right.

joelpittet commented on css/off.canvas.drawer.css:
```diff
@@ -0,0 +1,241 @@
+  font-family: Arial,Sans-Serif;
+  padding:10px;
+  font-size:14px;
+  line-height: 1.75em!important;
```
It may be good if we don't have to use `!important`, It's an id so it's pretty strong in specificity and harder to override in a child theme.

joelpittet commented on css/off.canvas.drawer.css:
```diff
@@ -0,0 +1,241 @@
+}
+
+.drawer-toggle--hamburger:hover {
+  -webkit-box-shadow: inset 1px 1px 1px #191938;
```
Should this be in the above statement on line 28? It seems like it would only work in webkit on hover.

joelpittet commented on css/off.canvas.drawer.css:
```diff
@@ -0,0 +1,241 @@
+#off-canvas-drawer--primary {
+  transition: transform 0.125s ease, visibility 0.125s ease;
+  visibility: visible;
+  height: 100% !important;
```
!important needed? May be good to look at all these?

joelpittet commented on templates/page/page.tpl.php:
```diff
@@ -132,6 +138,13 @@
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button style="float:right;margin-right:17px;" class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
+            <span>Menu ☰</span>
```
I like the hamburger icon but maybe it should be consistent with the other hamburger with `<span class="icon-bar"></span>`?

joelpittet commented on templates/page/page.tpl.php:
```diff
@@ -111,7 +117,7 @@
     <div class="span12">
       <div class="navbar">
-        <a class="btn btn-navbar" data-toggle="collapse" data-target="#ubc7-unit-navigation">
+        <a class="btn btn-navbar drawer-toggle--primary" data-toggle="collapse" data-target="#ubc7-unit-navigation">
```
Is this class needed since it's below?

joelpittet replied:
I'll try this out

joelpittet commented on templates/page/page.tpl.php:
```diff
@@ -132,6 +138,13 @@
       <?php print $fluidcontainerstart; ?>
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button style="float:right;margin-right:17px;" class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
```
The inline styles could be moved into the CSS file?

alice-0-kim commented:
@joelpittet I removed most of the important keywords and moved in-line styles to a CSS file. For the hamburger icon, did you want to keep the 'Menu' text besides the icon? Please let me know if further changes are required.

Thank you!

joelpittet commented:
The 'Menu' text is fine with me, maybe a question for @occupant but I'd say it could use a span and BEM class around the text to make it easier to hide/remove.

Thanks for fixing the styles @alice-0-kim 

Any thoughts on the last item? `drawer-toggle--primary` added in two places, is that needed?

alice-0-kim commented:
@joelpittet No worries at all. I'm assuming that you're referring to line 120 and 143 on the most updated `page.tpl.php`. I tried removing them, but it no longer triggers the action once `drawer-toggle--primary` gets removed. However, I might be looking at the wrong direction so please point me to the right direction.

joelpittet commented on template.php:
```diff
@@ -334,8 +334,44 @@ function megatron_preprocess_page(&$variables) {
+  }
+
+  // Add js and css for navigation sticky option
+  if (theme_get_setting('clf_sticky_option')) {
```
Maybe this sticky option can be a separate pull request or is it needed for this to work?

joelpittet commented on js/off.canvas.drawer.js:
```diff
@@ -0,0 +1,37 @@
+      }
+    }
+  }
+}(jQuery));
```
Closures are good but you aren't using jQuery, maybe that can just not pass params & args in.

joelpittet commented on theme-settings.php:
```diff
@@ -150,6 +136,64 @@ function megatron_form_system_theme_settings_alter(&$form, &$form_state) {
+
+  $form['clf_navigation_option']['clf_navigation_placement'] = array(
+    '#type' => 'select',
+    '#title' => t('Choose the type of primary navigation that should be used on this website.'),
```
This may be better to have as a separate option related to drawer only because the primary can be left there and have a drawer with the region there now. hmmm, thoughts?

joelpittet commented on templates/page/page.tpl.php:
```diff
@@ -132,6 +138,18 @@
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
+            <div>Menu</div>
```
Span tag would be better here and the text could be nice to have it configurable with that as the default in the appearance settings.

alice-0-kim commented:
@joelpittet I also think it is appropriate to have a separate pull request for adding sticky option. Also, should I wait to see if James has any input on the last comment you made about having a separate option related to drawer?

Thank you!

joelpittet commented:
@occupant may have better wording and ideas on the names for the options. My guess would be it may be nice to start with "Use drawer menu?" And that could show additional options like naming the drawer menu title, the animation style, etc.  The one thing we agreed on for sure was the ability to name the drawer menu toggle in the appearance options.

joelpittet commented on js/off.canvas.drawer.js:
```diff
@@ -0,0 +1,37 @@
+  Drupal.behaviors.megatronDrawer = {
+    attach: function(context, settings) {
+      // Primary drawer toggle.
+      var pushTrigger = document.getElementsByClassName("drawer-toggle--primary");
```
This won't work in IE8, I'm fine with that but may want to double check with James to see what they are supporting. 
https://caniuse.com/#search=getElementsByClassName

occupant replied:
I don't think that should be an issue - we don't support anything older than IE10 for new projects (unless someone wants to make that a requirement, in which case we would expect to make adjustments). I figure IE9 is a very reasonable baseline these days.

occupant replied:
I should have prefaced that with "Of course, this is just my opinion".

joelpittet replied:
Ok cool, this review can be closed.

occupant commented on css/off.canvas.drawer.css:
```diff
@@ -0,0 +1,263 @@
+  height:2px;
+  background-color:#000;
+  -webkit-border-radius:1px;
+  -moz-border-radius:1px;
```
Assuming IE9 is our baseline, we should also remove box-shadow prefixes 
https://caniuse.com/#search=box-shadow

occupant replied:
Also, it looks like the less source file isn't committed - @alice-0-kim would it be possible to add that?

alice-0-kim replied:
@occupant perhaps you are referring to the `less` folder? If not, please point me in the right direction.

occupant replied:
@alice-0-kim yes, that's right, the .less files in the less directory.

alice-0-kim replied:
@occupant I see! I just included `off.canvas.drawer.less` file to `less` folder, and it still needs a review on naming variables.

joelpittet replied:
The autoprefixer pull request can deal with this review in a more generic way:) Thanks again for the PR @alice-0-kim

occupant commented on theme-settings.php:
```diff
@@ -150,6 +136,57 @@ function megatron_form_system_theme_settings_alter(&$form, &$form_state) {
+    '#default_value' => theme_get_setting('clf_navigation_placement'),
+    '#options' => array(
+      'default' => t('Default CLF - Horizontal'),
+      'drawer--push-left' => t('Left Push Drawer'),
```
Maybe to make the labels clearer 

```
Default: no drawer region
Drawer Region: push from left
Drawer Region: cover from left
Drawer Region: push from right
Drawer Region: cover from right
```

And I guess the `clf_navigation_placement` setting is poorly named, since it's not just for navigation. Maybe something more like `clf_drawer_region`?
